blaze-client/src/main/scala/org/http4s/blaze/client/PoolManager.scala

```diff
@@ -173,14 +205,27 @@ private final class PoolManager[F[_], A <: Connection[F]](
         if (!isClosed) {
           def go(): F[Unit] =
             getConnectionFromQueue(key).flatMap {
-              case Some(conn) if !conn.isClosed =>
-                F.delay(logger.debug(s"Recycling connection for $key: $stats")) *>
-                  F.delay(callback(Right(NextConnection(conn, fresh = false))))
-
-              case Some(closedConn @ _) =>
-                F.delay(logger.debug(s"Evicting closed connection for $key: $stats")) *>
-                  decrConnection(key) *>
-                  go()
+              case Some(conn) =>
+                if (conn.isClosed) {
+                  F.delay(logger.debug(s"Evicting closed connection for $key: $stats")) *>
+                    decrConnection(key) *>
+                    go()
+                } else {
+                  conn.borrowDeadline match {
+                    case Some(deadline) if deadline.isOverdue() =>
+                      F.delay(
+                        logger.debug(
+                          s"Shutting down and evicting expired connection for $key: $stats"
+                        )
+                      ) *>
+                        decrConnection(key) *>
+                        F.delay(conn.shutdown()) *>
+                        go()
+                    case _ =>
+                      F.delay(logger.debug(s"Recycling connection for $key: $stats")) *>
+                        F.delay(callback(Right(NextConnection(conn, fresh = false))))
+                  }
+                }
```

Comment:
Perhaps a flatter outline? 
```suggestion
              case Some(conn) if conn.isClosed =>
                F.delay(logger.debug(s"Evicting closed connection for $key: $stats")) *>
                  decrConnection(key) *>
                  go()
              case Some(conn) if conn.borrowDeadline.exists(_.isOverdue()) => 
                F.delay(logger.debug(s"Shutting down and evicting expired connection for $key: $stats")) *>
                  decrConnection(key) *>
                  F.delay(conn.shutdown()) *>
                  go()
              case Some(conn) =>
                F.delay(logger.debug(s"Recycling connection for $key: $stats")) *>
                  F.delay(callback(Right(NextConnection(conn, fresh = false))))
              }
```

Reply:
Is scalac smart enough to do the Option match once and then the guards?

Compared to all the network stuff, checking the option two extra times really makes no difference anyway, and this is cleaner.